src/pybind/mgr/dashboard/services/rbd.py

```diff
@@ -427,8 +429,7 @@ def _rbd_pool_image_refs(cls, pool_names: List[str], namespace: Optional[str] =
                     # images without namespace
                     namespaces.append('')
                 for current_namespace in namespaces:
-                    ioctx.set_namespace(current_namespace)
-                    image_refs = cls._rbd_image_refs(ioctx)
+                    image_refs = cls._rbd_image_refs(ioctx, current_namespace)
```

Comment:
```
>>> i = c.open_ioctx('rbd_1')
>>> id(i)
140181373194872
>>> j = c.open_ioctx('rbd_1')
>>> id(j)
140181373194760
>>> i is j
False
>>> i == j
False
```

To achieve efficient caching, you probably need to cache the`iotctx`, and hence the `open_ioctx()` call too. The challenge with that is that when an ioctx is removed from the cache, `ioctx.close()` should be invoked.

I checked the Python librados code and `close()` is called on `__dealloc__()`, which I assume is invoked when refcount reaches 0. However, when you add caching to the equation you cannot fully trust the refcount mechanism, since:
* We are storing object refs as dictionary keys (for example the `_rbd_image_refs()` cache will store a reference to each ioctx in the dictionary key),
* We are storing object refs as dictionary values (in the suggested cached `open_ioctx()`, we would store a reference to each ioctx as a dict return value).

That's why many caches store [`weakrefs`](https://docs.python.org/3/library/weakref.html) instead.

Anyway, we may assume that we want ioctx to be long-living objects. However, the Dashboard `ttl_cache` doesn't perform any kind of garbage collection and only removes expired cache entries when the max cache size is hit.

So one solution would be to add a `mode` flag to the `ttl_cache` to enable a 2 modes of cache clean-ups:
* **Lazy cleanup** (current mode): where the stale/expired items are only removed when: TTL has expired or forced by the cache pressure (max number of cache items achieved).
* **Eager cleanup** (new mode): on every look-up the cache is inspected from the first/older item (we use an OrderedDict, so item 0 will be the oldest one) for expired items and they are progressively removed until a non-expired item is found.

---
A completely different approach is avoid caching stateful objects and go with identifiers (strings: pools, namespaces):
```py
def _rbd_pool_image_refs(cls, pool_names: List[str], namespace: Optional[str] = None):
  return [{'image_name': image, 'namespace': namespace, 'pool_name': pool}
      for pool in  pools
      for namespace in cached_function_that_returns_namespaces(pool)
      for image in cached_function_that_returns_images(pool, namespace)
    ]
```

Reply:
I dont think we want to store ioctx at all. I also believe that that the snippet you pasted would mean opening an ioctx twice. So my idea would be to have something like:

```
@ttl_cache(10, skip=['ioctx'])
def _rbd_image_refs(cls, ioctx, pool_name, namespace=''):
...


```


So that it doesn't cache the ioctx.

Reply:
Well now that I think about it there won't be much rbd pools so it shouldn't be a problem to open more than once without caching